Requireify of the uproxy repo

iislucas opened the pull request:
Moves the main uProxy repo to using require-style imports. This also includes many code cleanups, removal of dead code, and refactorings to remove some of the circular dependencies. 

TESTED: 
 * grunt test
 * Interactively loaded and tried the uProxy Chrome-App, Chrome-Extension, and Firefox Addon

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1318)
<!-- Reviewable:end -->

dborkan commented:
4,338 additions and 13,841 deletions - who won our bet?

dborkan commented:
Read through about half of it, will continue in the office later today...

---





<sup>**[src/chrome/app/scripts/chrome_oauth.ts, line 5 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRI3-mx-0JU43r2SCD)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/chrome/app/scripts/chrome_oauth.ts#L5)):</sup>
I don't understand why some imports use words_separated_by_underscore and others use AltCaps.  Typically we use AltCaps for class/module/enum names, and altCaps (starting with lowercase) for variable names.  Addding a 3rd type seems confusing, and it's especially confusing in the case of just one word (e.g. later on we have "import firewall = require...", looking at code later on in that file one would think "firewall" was an instance of some Firewall object, when in fact it's a module)

---

<sup>**[src/chrome/extension/manifest.json, line 38 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRHDwu9dTi2zM91-0q)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/chrome/extension/manifest.json#L38)):</sup>
Not really related to this pull request but it might be good to remove the options page here since it's kinda confusing if users click on it and just see the uProxy UI in a full tab

---

<sup>**[src/generic_core/freedom-module.ts, line 22 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRGTahh4OohPosmtf_)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/generic_core/freedom-module.ts#L22)):</sup>
Many of these import/require statements aren't used in this file (it looks like most of this content has been moved to upoxy_core.ts).  Can these be removed?

---

<sup>**[src/generic_core/freedom-module.ts, line 69 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRJireuaM9PGWhcK7F)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/generic_core/freedom-module.ts#L69)):</sup>
Feel free to remove these commented out TODO onCommands.  They've been here forever and I don't think we actually want to do them anymore (or if we do invites, we don't necessarily need to do them this way)

---

<sup>**[src/generic_core/storage.ts, line 17 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRLANodks1R2sOtW0_)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/generic_core/storage.ts#L17)):</sup>
Can we double check that this is really what we want?  I believe freedom['storage'] and freedom['core.storage'] are actually 2 different implementations of storage.  If so we need to do a lot of testing to make sure this still works well, and also this change is backwards compatible (i.e. our closed-beta-testers won't lose all their settings when they upgrade to the newest uProxy)

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

iislucas commented:
<sup>**[src/chrome/app/scripts/chrome_oauth.ts, line 5 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRI3-mx-0JU43r2SCD)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/chrome/app/scripts/chrome_oauth.ts#L5)):</sup>
modules are lower_case_underscore_separated, ClassesAreCamlCase

---

<sup>**[src/chrome/extension/manifest.json, line 38 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRHDwu9dTi2zM91-0q)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/chrome/extension/manifest.json#L38)):</sup>
Done.

---

<sup>**[src/generic_core/freedom-module.ts, line 22 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRGTahh4OohPosmtf_)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/generic_core/freedom-module.ts#L22)):</sup>
Done.

---

<sup>**[src/generic_core/freedom-module.ts, line 69 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRJireuaM9PGWhcK7F)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/generic_core/freedom-module.ts#L69)):</sup>
Done.

---

<sup>**[src/generic_core/storage.ts, line 17 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRLANodks1R2sOtW0_)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/generic_core/storage.ts#L17)):</sup>
I asked Ray and he told me they were the same in this case... @ryscheng, did I understand this correctly? 

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

dborkan commented:
<sup>**[src/chrome/app/scripts/chrome_oauth.ts, line 5 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRI3-mx-0JU43r2SCD)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/chrome/app/scripts/chrome_oauth.ts#L5)):</sup>
Modules used to be AltCaps, why change this for requireify?  Also see my comment on how this is confusing for one world modules, e.g. firewall, which now aren't easily distinguished from variables (like an instance of a Firewall object)...  Also why add one more rule to the style guide?  Now uProxy code has altCaps, AltCaps, and under_score, I could imagine most people taking a quick look at the code will just think we have no rules.

---

<sup>**[src/generic_ui/polymer/context.d.ts, line 4 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRaDHIUOC2ZHNPHm5X)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/polymer/context.d.ts#L4)):</sup>
Can we come up with a better name for this?  StaticInPanel doesn't really mean anything to me, especially the word "static" here is vague.

---

<sup>**[src/generic_ui/polymer/context.d.ts, line 10 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRaOTbk_fm3OG9AjKE)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/polymer/context.d.ts#L10)):</sup>
Can we come up with a better name for this too?   I don't like including "browserified" in the name.  Polymer code which uses this shouldn't care that browserify was used.  Even just "globals" or "ui_globals" would be better, or something indicating that all of these are bound to a background page, like "background_exports" or "background_vars", etc

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

dborkan commented:
<sup>**[src/generic_ui/polymer/context.d.ts, line 10 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRbkkeMkS1mpsZkmZU)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/polymer/context.d.ts#L10)):</sup>
Also why do we even need to combine these 3 into one global structure?  Why not just have 3 global variables like we've done in the past?  That way we don't need to change all "ui" to "browserified_exports.ui"?

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

dborkan commented:
<sup>**[src/generic_ui/scripts/ui.ts, line 16 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRdB7FNd5POZxTCF9l)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/scripts/ui.ts#L16)):</sup>
Is it possible we could get rid of the ```import user = require('./user');``` and just change these 2 lines to something like ```import User = require('./user').User;``` (not sure the exact syntax).  The problem right now is we import "user", but "user" is also used as a variable name later in this file (line 322), which will result in shadowing this global "user" variable

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

dborkan commented:
<sup>**[src/chrome/app/background.html, line 4 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRe-ijjUhErtfRkokn)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/chrome/app/background.html#L4)):</sup>
Why is it that here we have to include lodash.min.js from the HTML file, while in the UI we are able to include it using require (generic_ui/scripts/user.ts)?

---

<sup>**[src/interfaces/social.ts, line 208 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRfIsCMs9G04AkOBuP)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/interfaces/social.ts#L208)):</sup>
I'm pretty confused by "RemoteUser" and some of the similar interfaces here.  Why is this needed?  Why does each network in this file have a roster that is a map from a string to a RemoteUser, when the actual network in JavaScript has a mapping from a string to a "User" object?

Related to this, why in uproxy_core.ts does getUser claim to return a RemoteUser object, when really it always returns a User object?

Is it expected that any time we add a new public method to User, we add the same method definition to RemoteUser?

Is it possible to just get rid of this so we don't have to maintain the type in 2 places?

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

iislucas commented:
<sup>**[src/chrome/app/background.html, line 4 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRe-ijjUhErtfRkokn)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/chrome/app/background.html#L4)):</sup>
When we moved to npm, this was accidentally left in. Removed now. 

---

<sup>**[src/chrome/app/scripts/chrome_oauth.ts, line 5 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRI3-mx-0JU43r2SCD)** ([raw file](https://github.com/uproxy/uproxy/blob/18d667f029188fe986fe37868f3db66f48e62072/src/chrome/app/scripts/chrome_oauth.ts#L5)):</sup>
in uproxy networking and uproxy lib they were always lower case. Using require-style means we had a lot more modules, which is great because we now know where things come from, so this became even more important to distinguish. I did initially try AltCaps for module names, but this was much messier. So I think this is the right thing to do. 

---

<sup>**[src/generic_ui/polymer/context.d.ts, line 4 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRaDHIUOC2ZHNPHm5X)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/polymer/context.d.ts#L4)):</sup>
`Globals` ?

---

<sup>**[src/generic_ui/polymer/context.d.ts, line 10 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRaOTbk_fm3OG9AjKE)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/polymer/context.d.ts#L10)):</sup>
not sure; background_exports has the advantage that it avoids clashing with other things called global/globals etc.  

---

<sup>**[src/generic_ui/polymer/context.d.ts, line 10 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRbkkeMkS1mpsZkmZU)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/polymer/context.d.ts#L10)):</sup>
Because browserify only exports a single variable. 

---

<sup>**[src/generic_ui/scripts/ui.ts, line 16 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRdB7FNd5POZxTCF9l)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/scripts/ui.ts#L16)):</sup>
nope; that's not supported in TS. 

---

<sup>**[src/interfaces/social.ts, line 208 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRfIsCMs9G04AkOBuP)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/interfaces/social.ts#L208)):</sup>
There's two factors, primarily: the set of interfaces was and still is a mess; rather than do a more significant cleanup, we did the relatively direct thing. Secondarily: it's best to avoid circular definitions, and one of the tools to do that is to separate your abstract interface from your implementation. 

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

dborkan commented:
I discussed this with Lucas in person yesterday.  We want to get rid of as much of the browserified_exports.{ui,model,core} as possible before merging (or in the worst case just change browserified_exports to a simpler shorter name).  Once that's done this is good to merge

---





<sup>**[src/generic_ui/scripts/ui.ts, line 17 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRdB7FNd5POZxTCF9l)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/scripts/ui.ts#L17)):</sup>
Maybe we should change the ```import user = require('./user')``` line to be something like ```import user_module = require('./user')```?  I don't think the user module is used for anything other than to import User

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

willscott commented:
Very excited to see this saga finally reaching completion! :tada:

dborkan commented:
:+1: yay let's get it into dev!

jpevarnek commented:
:fireworks: :fireworks:

iislucas commented:
<sup>**[src/generic_ui/scripts/ui.ts, line 17 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1318#-JnRdB7FNd5POZxTCF9l)** ([raw file](https://github.com/uproxy/uproxy/blob/64b71f4cfb6bb1d08d5a4b0575fdcb12a26fbf60/src/generic_ui/scripts/ui.ts#L17)):</sup>
Done.

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1318)
<!-- Sent from Reviewable.io -->

iislucas commented:
Got verbal LGTM from @dborkan merging time!
